leapp/reporting/__init__.py

```diff
@@ -5,102 +5,256 @@
 
-    detail = fields.JSON()
-    """
-    Detail of the report entry as JSON data
-    """
+    @property
+    def value(self):
+        """Return report field value"""
+        return self._value
 
-    renderers = fields.Model(Renderers)
-    """
-    :class:`Renderers` describe how to render this report entry
-    """
+    @property
+    def path(self):
+        """Return report field destination path (nested dict description)"""
+        raise NotImplementedError
 
-    audience = fields.List(fields.StringEnum(choices=['developer', 'sysadmin']))
-    """
-    Who is the main audience of this report entry
-    """
+    def apply(self, report):
+        """Add report entry to the report dict based on provided path"""
+        _add_to_dict(report, self.path, self.value, self.is_list)
 
-    flags = fields.List(fields.StringEnum(choices=['inhibitor']))
-    """
-    Flags can give a special meaning to each report entry, currently supported flags:
 
-        inhibitor - This report entry will inhibit the upgrade process
-    """
+class Title(BasePrimitive):
+    """Report title"""
+    @property
+    def path(self):
+        return ('title',)
+
+
+class Summary(BasePrimitive):
+    """Report summary"""
+    @property
+    def path(self):
+        return ('summary',)
+
+
+class Severity(BasePrimitive):
+    """Report severity"""
+
+    INFO = 'info'
+    LOW = 'low'
+    MEDIUM = 'medium'
+    HIGH = 'high'
+
+    def __init__(self, value=None):
+        if not isinstance(value, str):
+            raise TypeError('Value of "{}" must be a string'.format(self.__class__.__name__))
+        severities = ('info', 'low', 'medium', 'high')
+        if value not in severities:
+            raise ValueError('Severity should be one of: "{}"'.format(' '.join(severities)))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('severity',)
+
+
+class Audience(BasePrimitive):
+    """Target audience of the report"""
+    def __init__(self, value=None):
+        if not isinstance(value, str):
+            raise TypeError('Value of "{}" must be a string'.format(self.__class__.__name__))
+        audiences = ('sysadmin', 'developer')
+        if value not in audiences:
+            raise ValueError('Audience should be one of: "{}"'.format(' '.join(audiences)))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('audience',)
+
+
+class Flags(BasePrimitive):
+    """Report flags"""
 
+    INHIBITOR = 'inhibitor'
+    FAILURE = 'failure'
 
-def report(title=None, detail=None, renderers=None, audience=None, flags=None, severity=None):
+    def __init__(self, value=None):
+        if not isinstance(value, list):
+            raise TypeError('Value of "{}" must be a list'.format(self.__class__.__name__))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('flags',)
+
+
+class Tags(BasePrimitive):
+    """Report tags"""
+
+    ACCESSIBILITY = 'accessibility'
```

Comment:
How will those tags be used? And who will be responsible for adding a new tag if necessary?

Reply:
https://github.com/oamg/leapp-repository/blob/3c8c66eaef327c5c5eaa4e28d60aa07931d6caee/repos/system_upgrade/el7toel8/actors/authselectapply/actor.py#L40

Reply:
Mostly reviewers I would say but we will have to document it properly.

Reply:
For validation purposes I would change this to something else, but this is _if_ you do NOT want to support other values than listed here. Which is I feel like it is the idea, no?

```python
class Tags(BasePrimitive):
    """Report tags"""
    class _Value(object):
        def __init__(self, value):
            self.value = value

    ACCESSIBILITY = Tags._Value('accessibility')
    ....
    def __init__(self, value=None):
        if not isinstance(value, list):
            raise TypeError('Value of "Tags" must be a list')
        if not all([isinstance(v, Tags._Value) for v in value]):
            raise TypeError('Unsupported tag value passed for Report Tags.')
        self._value = value
```

Reply:
Wonderful. Yep, that is what I wanted.

You probably meant:

`ACCESSIBILITY = _Value('accessibility')`

and not 

`ACCESSIBILITY = Tags._Value('accessibility')`

right?

Reply:
TBH i wasn't sure if just _Value works, but it should IIRC

Reply:
implemented